### PR TITLE
[Resolve #1114] Resolvable Template Handler configs and the !stack_attr resolver

### PR DESCRIPTION
**Pull request by jfalkenstein**
This pull request adds two related things:
1. Template handler configurations are now fully resolvable.
2. You can now use the `!stack_attr` resolver to reference other configurations on the stack and StackGroup 

The need for these two things came about in my testing on the recent changes to Sceptre's resolvers. 

## Why is this necessary?
Previously, I had this on one of my stacks:

```yaml
template:
    type: sam
    path: beatbox/template.yaml
    artifact_bucket_name: {{ template_bucket_name }}
    artifact_prefix: {{ template_key_prefix }}
```

^^ This has been working just fine until I set `template_bucket_name` in my project to a `!stack_output` resolver. Once I did that, however, the template handler blew up because Jinja was attempting to cast the StackOutput resolver to string.

With the changes in this PR, I can now do this:

```yaml
template:
    type: sam
    path: beatbox/template.yaml
    # This will access and resolve the value of template_bucket_name from the StackGroup 
    # and then access that value as the resolved value for handler.
    artifact_bucket_name: !stack_attr template_bucket_name
    artifact_prefix: {{ template_key_prefix }}
```

## PR Checklist

- [x] Wrote a good commit message & description [see guide below].
- [x] Commit message starts with `[Resolve #issue-number]`.
- [x] Added/Updated unit tests.
- [x] Added/Updated integration tests (if applicable).
- [x] All unit tests (`make test`) are passing.
- [x] Used the same coding conventions as the rest of the project.
- [x] The new code passes pre-commit validations (`pre-commit run --all-files`).
- [x] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

### CODE REVIEW

File: docs/_source/docs/resolvers.rst
Comment (zaro0508) [inline docs/_source/docs/resolvers.rst:111]: does ordering matter?  can you put paramters above sceptre_user_data?

Would it be possible to use more concrete examples for `key` and the values for it?   Also could we add a comment for `my_parameter` telling users what  `!stack_attr sceptre_user_data.key.1` will pass to my_parameter?

File: docs/_source/docs/resolvers.rst
Comment (zaro0508) [inline docs/_source/docs/resolvers.rst:98]: I think users would expect that  `artifact_bucket_name: {{ stack_group_config.template_bucket_name }}` would work as well.  if it doesn't then can you please provide some explanation as to why it would not work?

**Review by zaro0508 — CHANGES_REQUESTED** [review]

File: docs/_source/docs/resolvers.rst
Comment (zaro0508) [reply docs/_source/docs/resolvers.rst:98]: does the template handler doc need updating? https://github.com/Sceptre/sceptre-sam-handler#arguments section.. 
```
artifact_bucket_name (string, required): The bucket name where artifacts should be uploaded to on S3 during the packaging process. If your project has a template_bucket_name, you can set this to {{ template_bucket_name }}.
```

Comment (jfalkenstein) [reply docs/_source/docs/resolvers.rst:98]: > I think users would expect that artifact_bucket_name: {{ stack_group_config.template_bucket_name }} would work as well

The problem here is that template_bucket_name (and many other items that might be in stack_group_config) might be resolvers. Especially with the recent set of merges to master, there are a lot more resolvable properties and "inheriting" a resolver is a lot more likely. Remember, the order of resolution within Sceptre is:

1. Gather raw stack group config variables from higher-level stack group config.yaml files. Since the yaml resolver constructs the resolvers, these variables might not be strings but instead might be raw resolvers.
2. (**"Render time"**) Use those stack group config and user variables to _render_ the config file using Jinja into a string. Thus `{{ stack_group_config.template_bucket_name }}` would cast the resolver to string rather than resolve it.
3. Parse the string into yaml, which will instantiate resolvers on the Stack.
4. Use that yaml to create the Stack instance, which assigns those resolvers to resolvable properties
5. (**"Resolve time"**) Then, upon accessing those resolvable properties, those resolvers are resolved into values.

Because of this, you cannot access resolved values at render time (step 2 above). This is the whole reason why I made the `!stack_attr` resolver, because we need to access other stack attributes (which are set at step 4 and resolved at step 4). Simply accessing the pre-resolved values at render time doesn't work. 

> if it doesn't then can you please provide some explanation as to why it would not work?

I hope the above works. However, I was thinking that one thing that hasn't been documented in Sceptre's docs is the "Stack construction lifecycle", similar to what I just wrote above. I'll add a docs page that explains this, since I find it's frequently confused in the conversations I've had with you and other maintainers lately.

> does the template handler doc need updating? https://github.com/Sceptre/sceptre-sam-handler#arguments section.

Yeah, it will. However, I didn't want to update it until this change has been released as a version. Otherwise, I'd be documenting functionality that doesn't exist yet. The extension repos make more sense if they "chase" Sceptre versions rather than attempt to preempt them.

File: docs/_source/docs/resolvers.rst
Comment (jfalkenstein) [reply docs/_source/docs/resolvers.rst:111]: > does ordering matter? can you put parameters above sceptre_user_data?

Sure? I'm not sure why it matters. Remember render time is prior to resolve time, so the order they appear in on the template doesn't really matter. In this particular case, I wanted to introduce the data we're selecting first before selecting it. 

> Would it be possible to use more concrete examples for key and the values for it? Also could we add a comment for my_parameter telling users what !stack_attr sceptre_user_data.key.1 will pass to my_parameter?

Yeah, I could definitely do that.

**Comment by jfalkenstein** [comment]
@zaro0508 I've added more examples and a new documentation section on the resolution order of values on a StackGroup config. Hopefully this clears everything up for you.

File: tests/test_resolvers/test_cache.py
Comment (jfalkenstein) [inline tests/test_resolvers/test_cache.py:1]: I found these tests. Somebody committed these and I'm not sure why they still exist!

**Comment by ykhalyavin** [comment]
Honestly, it feels like this resolver adds lots of ambiguity and resolving resolvable properties should be a part of sceptre. E.g. at render time sceptre could check if a stack_group_config is resolvable property, then resolve it and substitute the resolved value transparently to the user.

**Comment by jfalkenstein** [comment]
@ykhalyavin What you suggest is not possible. As I documented on resolution order, resolvers require a Stack and the Stack doesn't exist at render time.

File: docs/_source/docs/resolvers.rst
Comment (zaro0508) [reply docs/_source/docs/resolvers.rst:98]: > The problem here is that template_bucket_name (and many other items that might be in stack_group_config) might be resolvers.

What if `template_bucket_name` is not a resolver?  I have template_bucket_name set to a static value in my StackGroupConfig (config.yaml) files.  Should I not be able to use {{ stack_group_config.template_bucket_name }} in my stack config files in this case?  I'm just saying that is is documenting for a specific case and you should make it clear why `!stack_attr` needs to be used instead of `{{ stack_group_config.template_bucket_name }}`

File: docs/_source/docs/stack_config.rst
Comment (zaro0508) [inline docs/_source/docs/stack_config.rst:366]: thanks for adding this info to the doc.

File: docs/_source/docs/stack_config.rst
Comment (zaro0508) [inline docs/_source/docs/stack_config.rst:405]: i think this is the reason for using `!stack_attr` resolver?  if so maybe tie this info back to that resolver and vice versa?

**Review by zaro0508 — CHANGES_REQUESTED** [review]

File: docs/_source/docs/resolvers.rst
Comment (jfalkenstein) [reply docs/_source/docs/resolvers.rst:98]: Great point! I'll make that change.

File: docs/_source/docs/stack_config.rst
Comment (jfalkenstein) [reply docs/_source/docs/stack_config.rst:405]: Yeah, I can do that. Thanks for the input!

File: docs/_source/conf.py
Comment (jfalkenstein) [inline docs/_source/conf.py:214]: There were some warnings being output by Sphinx about it not knowing how to resolve docs for these classes.

File: docs/_source/docs/permissions.rst
Comment (jfalkenstein) [inline docs/_source/docs/permissions.rst:128]: "on non-cloudformation actions" made no sense in this context. I fixed that.

**Comment by jfalkenstein** [comment]
Alright, @zaro0508, I've refined the docs some, making the changes you've requested.

**Review by zxiiro — APPROVED** [review]

**Review by zxiiro — COMMENTED** [review]
Sorry miss-clicked approved earlier when I meant to leave a comment.

File: docs/_source/docs/permissions.rst
Comment (jfalkenstein) [reply docs/_source/docs/permissions.rst:130]: Yeah, I can do that.

File: docs/_source/docs/resolvers.rst
Comment (zaro0508) [inline docs/_source/docs/resolvers.rst:126]: is it possible to reference `stack_name`?  something like this?
```
parameters:
  ThisStack: !stack_attr stack_name
```

i'm wondering if it could be another work around for https://github.com/Sceptre/sceptre/discussions/1021

**Review by zaro0508 — APPROVED** [review]

File: docs/_source/docs/resolvers.rst
Comment (jfalkenstein) [reply docs/_source/docs/resolvers.rst:126]: Yes! That would work perfectly.

**Pull request merged**
